File: pkg/notifications/emergency_detector.go
package notifications

import (
	"time"

	"github.com/markus-lassfolk/autonomy/pkg/logx"
)

// EmergencyDetector detects emergency conditions that require immediate attention
type EmergencyDetector struct {
	config *IntelligenceConfig
	logger *logx.Logger
}

// NewEmergencyDetector creates a new emergency detector
func NewEmergencyDetector(config *IntelligenceConfig, logger *logx.Logger) *EmergencyDetector {
	return &EmergencyDetector{
		config: config,
		logger: logger,
	}
}

// DetectEmergency analyzes system state and alert context to detect emergency conditions
func (ed *EmergencyDetector) DetectEmergency(
	systemState *SystemState,
	alertType AlertType,
	baseData map[string]interface{},
) EmergencyLevel {
	if !ed.config.EmergencyDetectionEnabled {
		return EmergencyNone
	}

	maxLevel := EmergencyNone

	// Check system health emergencies
	if systemLevel := ed.checkSystemHealthEmergency(systemState.SystemHealth); systemLevel > maxLevel {
		maxLevel = systemLevel
	}

	// Check network health emergencies
	if networkLevel := ed.checkNetworkHealthEmergency(systemState.NetworkHealth); networkLevel > maxLevel {
		maxLevel = networkLevel
	}

	// Check cascading failure patterns
	if cascadeLevel := ed.checkCascadingFailures(systemState.ActiveIncidents, systemState.RecentFailures); cascadeLevel > maxLevel {
		maxLevel = cascadeLevel
	}

	// Check alert-specific emergency conditions
	if alertLevel := ed.checkAlertSpecificEmergency(alertType, baseData); alertLevel > maxLevel {
		maxLevel = alertLevel
	}

	// Check temporal emergency patterns
	if temporalLevel := ed.checkTemporalEmergency(systemState); temporalLevel > maxLevel {
		maxLevel = temporalLevel
	}

	if maxLevel > EmergencyNone {
		ed.logger.Warn("Emergency condition detected",
			"level", maxLevel,
			"alert_type", alertType,
			"system_cpu", systemState.SystemHealth.CPUUsage,
			"system_memory", systemState.SystemHealth.MemoryUsage,
			"system_temperature", systemState.SystemHealth.Temperature)
	}

	return maxLevel
}

// checkSystemHealthEmergency checks for system health emergencies
func (ed *EmergencyDetector) checkSystemHealthEmergency(health *SystemHealthState) EmergencyLevel {
	if health == nil {
		return EmergencyNone
	}

	thresholds := ed.config.EmergencyThresholds

	// Critical system conditions
	if health.CPUUsage >= thresholds.CPUUsageEmergency ||
		health.MemoryUsage >= thresholds.MemoryUsageEmergency ||
		health.Temperature >= thresholds.TemperatureEmergency ||
		health.DiskUsage >= thresholds.DiskUsageEmergency {
		return EmergencyCritical
	}

	// High emergency conditions (80% of critical thresholds)
	if health.CPUUsage >= thresholds.CPUUsageEmergency*0.8 ||
		health.MemoryUsage >= thresholds.MemoryUsageEmergency*0.8 ||
		health.Temperature >= thresholds.TemperatureEmergency*0.8 ||
		health.DiskUsage >= thresholds.DiskUsageEmergency*0.8 {
		return EmergencyHigh
	}

	// Medium emergency conditions (60% of critical thresholds)
	if health.CPUUsage >= thresholds.CPUUsageEmergency*0.6 ||
		health.MemoryUsage >= thresholds.MemoryUsageEmergency*0.6 ||
		health.Temperature >= thresholds.TemperatureEmergency*0.6 ||
		health.DiskUsage >= thresholds.DiskUsageEmergency*0.6 {
		return EmergencyMedium
	}

	return EmergencyNone
}

// checkNetworkHealthEmergency checks for network health emergencies
func (ed *EmergencyDetector) checkNetworkHealthEmergency(health *NetworkHealthState) EmergencyLevel {
	if health == nil {
		return EmergencyNone
	}

	thresholds := ed.config.EmergencyThresholds

	// Critical network conditions
	if !health.PrimaryInterfaceUp && health.BackupInterfacesUp == 0 {
		return EmergencyCritical // Complete network failure
	}

	if health.AveragePacketLoss >= thresholds.PacketLossEmergency ||
		health.AverageLatency >= thresholds.LatencyEmergency {
		return EmergencyCritical
	}

	// High emergency conditions
	if !health.PrimaryInterfaceUp ||
		health.BackupInterfacesUp < health.TotalInterfaces/2 {
		return EmergencyHigh
	}

	if health.AveragePacketLoss >= thresholds.PacketLossEmergency*0.6 ||
		health.AverageLatency >= thresholds.LatencyEmergency*0.6 {
		return EmergencyHigh
	}

	// Medium emergency conditions
	if health.AveragePacketLoss >= thresholds.PacketLossEmergency*0.3 ||
		health.AverageLatency >= thresholds.LatencyEmergency*0.3 {
		return EmergencyMedium
	}

	return EmergencyNone
}

// checkCascadingFailures checks for cascading failure patterns
func (ed *EmergencyDetector) checkCascadingFailures(incidents []ActiveIncident, failures []FailureRecord) EmergencyLevel {
	thresholds := ed.config.EmergencyThresholds
	now := time.Now()

	// Count recent high-severity incidents
	recentHighSeverity := 0
	for _, incident := range incidents {
		if incident.Severity >= int(EmergencyHigh) &&
			now.Sub(incident.StartTime) < 10*time.Minute {
			recentHighSeverity++
		}
	}

	if recentHighSeverity >= thresholds.CascadingFailureCount {
		return EmergencyCritical
	}

	// Count failures in the last minute
	recentFailures := 0
	cutoff := now.Add(-1 * time.Minute)
	for _, failure := range failures {
		if failure.Timestamp.After(cutoff) {
			recentFailures++
		}
	}

	failureRate := float64(recentFailures)
	if failureRate >= thresholds.FailureRateEmergency {
		return EmergencyHigh
	}

	if failureRate >= thresholds.FailureRateEmergency*0.6 {
		return EmergencyMedium
	}

	return EmergencyNone
}

// checkAlertSpecificEmergency checks for alert-type specific emergency conditions
func (ed *EmergencyDetector) checkAlertSpecificEmergency(alertType AlertType, baseData map[string]interface{}) EmergencyLevel {
	switch alertType {
	case AlertThermal:
		if temp, ok := baseData["temperature"].(float64); ok {
			if temp >= ed.config.EmergencyThresholds.TemperatureEmergency {
				return EmergencyCritical
			}
			if temp >= ed.config.EmergencyThresholds.TemperatureEmergency*0.8 {
				return EmergencyHigh
			}
		}

	case AlertFailover:
		// Multiple rapid failovers indicate emergency
		if failoverCount, ok := baseData["recent_failover_count"].(int); ok {
			if failoverCount >= 3 {
				return EmergencyHigh
			}
		}

	case AlertConnectivityIssue:
		if loss, ok := baseData["packet_loss"].(float64); ok {
			if loss >= ed.config.EmergencyThresholds.PacketLossEmergency {
				return EmergencyCritical
			}
		}
		if latency, ok := baseData["latency"].(float64); ok {
			if latency >= ed.config.EmergencyThresholds.LatencyEmergency {
				return EmergencyHigh
			}
		}

	case AlertSystemHealth:
		if severity, ok := baseData["severity"].(string); ok {
			switch severity {
			case "critical":
				return EmergencyCritical
			case "high":
				return EmergencyHigh
			case "medium":
				return EmergencyMedium
			}
		}

	case AlertDataLimit:
		if usage, ok := baseData["usage_percent"].(float64); ok {
			if usage >= 100.0 {
				return EmergencyMedium // Data limit exceeded
			}
		}

	case AlertPredictive:
		if confidence, ok := baseData["confidence"].(float64); ok {
			if confidence >= 90.0 {
				return EmergencyHigh // High confidence prediction
			}
			if confidence >= 70.0 {
				return EmergencyMedium
			}
		}
	}

	// Check for explicit emergency flag
	if emergency, ok := baseData["emergency"].(bool); ok && emergency {
		return EmergencyHigh
	}

	return EmergencyNone
}

// checkTemporalEmergency checks for time-based emergency patterns
func (ed *EmergencyDetector) checkTemporalEmergency(systemState *SystemState) EmergencyLevel {
	now := time.Now()

	// Check for prolonged incidents
	for _, incident := range systemState.ActiveIncidents {
		duration := now.Sub(incident.StartTime)

		if duration >= ed.config.EmergencyThresholds.ServiceDowntimeEmergency {
			if incident.Severity >= int(EmergencyHigh) {
				return EmergencyCritical
			} else {
				return EmergencyHigh
			}
		}
	}

	// Check for repeated failures of the same component
	componentFailures := make(map[string]int)
	cutoff := now.Add(-1 * time.Hour)

	for _, failure := range systemState.RecentFailures {
		if failure.Timestamp.After(cutoff) {
			componentFailures[failure.Component]++
		}
	}

	for component, count := range componentFailures {
		if count >= 5 { // 5 failures in an hour
			ed.logger.Warn("Repeated component failures detected",
				"component", component,
				"failures", count)
			return EmergencyMedium
		}
	}

	return EmergencyNone
}

// IsEmergencyLevel checks if a given level qualifies as an emergency
func IsEmergencyLevel(level EmergencyLevel) bool {
	return level > EmergencyNone
}

// GetEmergencyLevelString returns string representation of emergency level
func GetEmergencyLevelString(level EmergencyLevel) string {
	switch level {
	case EmergencyNone:
		return "none"
	case EmergencyLow:
		return "low"
	case EmergencyMedium:
		return "medium"
	case EmergencyHigh:
		return "high"
	case EmergencyCritical:
		return "critical"
	default:
		return "unknown"
	}
}

// GetEmergencyLevelPriority returns notification priority for emergency level
func GetEmergencyLevelPriority(level EmergencyLevel) int {
	switch level {
	case EmergencyNone:
		return PriorityNormal
	case EmergencyLow:
		return PriorityNormal
	case EmergencyMedium:
		return PriorityHigh
	case EmergencyHigh:
		return PriorityEmergency
	case EmergencyCritical:
		return PriorityEmergency
	default:
		return PriorityNormal
	}
}

File: src/c/autonomy-daemon/notifications/emergency_detector.c
#include "emergency_detector.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <time.h>
#include <math.h>

// Global emergency detector instance
static emergency_detector_t g_emergency_detector;
static bool g_emergency_detector_initialized = false;

// Initialize emergency detector
static int emergency_detector_init(const emergency_detector_config_t* config) {
    if (g_emergency_detector_initialized) {
        return 0; // Already initialized
    }
    
    if (!config) {
        return -1;
    }
    
    memset(&g_emergency_detector, 0, sizeof(emergency_detector_t));
    
    // Copy configuration
    g_emergency_detector.config = *config;
    
    // Initialize mutex
    g_emergency_detector.mutex = malloc(sizeof(pthread_mutex_t));
    if (!g_emergency_detector.mutex) {
        return -1;
    }
    
    pthread_mutex_init(g_emergency_detector.mutex, NULL);
    
    // Initialize incident tracking
    g_emergency_detector.active_incidents = malloc(config->max_active_incidents * sizeof(active_incident_t));
    if (!g_emergency_detector.active_incidents) {
        pthread_mutex_destroy(g_emergency_detector.mutex);
        free(g_emergency_detector.mutex);
        return -1;
    }
    
    g_emergency_detector.max_active_incidents = config->max_active_incidents;
    g_emergency_detector.active_incidents_count = 0;
    
    // Initialize failure records
    g_emergency_detector.recent_failures = malloc(config->max_failure_records * sizeof(failure_record_t));
    if (!g_emergency_detector.recent_failures) {
        free(g_emergency_detector.active_incidents);
        pthread_mutex_destroy(g_emergency_detector.mutex);
        free(g_emergency_detector.mutex);
        return -1;
    }
    
    g_emergency_detector.max_failure_records = config->max_failure_records;
    g_emergency_detector.failure_records_count = 0;
    
    g_emergency_detector_initialized = true;
    return 0;
}

// Clean up emergency detector
static void emergency_detector_cleanup(void) {
    if (!g_emergency_detector_initialized) return;
    
    if (g_emergency_detector.mutex) {
        pthread_mutex_destroy(g_emergency_detector.mutex);
        free(g_emergency_detector.mutex);
    }
    
    if (g_emergency_detector.active_incidents) {
        free(g_emergency_detector.active_incidents);
    }
    
    if (g_emergency_detector.recent_failures) {
        free(g_emergency_detector.recent_failures);
    }
    
    g_emergency_detector.active_incidents = NULL;
    g_emergency_detector.recent_failures = NULL;
    g_emergency_detector.mutex = NULL;
    g_emergency_detector.active_incidents_count = 0;
    g_emergency_detector.max_active_incidents = 0;
    g_emergency_detector.failure_records_count = 0;
    g_emergency_detector.max_failure_records = 0;
    
    g_emergency_detector_initialized = false;
}

// Check system health emergency
static emergency_level_t check_system_health_emergency(const system_health_state_t* health) {
    if (!health || !g_emergency_detector_initialized) {
        return EMERGENCY_LEVEL_NONE;
    }
    
    const emergency_thresholds_t* thresholds = &g_emergency_detector.config.emergency_thresholds;
    
    // Critical system conditions
    if (health->cpu_usage >= thresholds->cpu_usage_emergency ||
        health->memory_usage >= thresholds->memory_usage_emergency ||
        health->temperature >= thresholds->temperature_emergency ||
        health->disk_usage >= thresholds->disk_usage_emergency) {
        return EMERGENCY_LEVEL_CRITICAL;
    }
    
    // High emergency conditions (80% of critical thresholds)
    if (health->cpu_usage >= thresholds->cpu_usage_emergency * 0.8 ||
        health->memory_usage >= thresholds->memory_usage_emergency * 0.8 ||
        health->temperature >= thresholds->temperature_emergency * 0.8 ||
        health->disk_usage >= thresholds->disk_usage_emergency * 0.8) {
        return EMERGENCY_LEVEL_HIGH;
    }
    
    // Medium emergency conditions (60% of critical thresholds)
    if (health->cpu_usage >= thresholds->cpu_usage_emergency * 0.6 ||
        health->memory_usage >= thresholds->memory_usage_emergency * 0.6 ||
        health->temperature >= thresholds->temperature_emergency * 0.6 ||
        health->disk_usage >= thresholds->disk_usage_emergency * 0.6) {
        return EMERGENCY_LEVEL_MEDIUM;
    }
    
    return EMERGENCY_LEVEL_NONE;
}

// Check network health emergency
static emergency_level_t check_network_health_emergency(const network_health_state_t* health) {
    if (!health || !g_emergency_detector_initialized) {
        return EMERGENCY_LEVEL_NONE;
    }
    
    const emergency_thresholds_t* thresholds = &g_emergency_detector.config.emergency_thresholds;
    
    // Critical network conditions
    if (!health->primary_interface_up && health->backup_interfaces_up == 0) {
        return EMERGENCY_LEVEL_CRITICAL; // Complete network failure
    }
    
    if (health->average_packet_loss >= thresholds->packet_loss_emergency ||
        health->average_latency >= thresholds->latency_emergency) {
        return EMERGENCY_LEVEL_CRITICAL;
    }
    
    // High emergency conditions
    if (!health->primary_interface_up ||
        health->backup_interfaces_up < health->total_interfaces / 2) {
        return EMERGENCY_LEVEL_HIGH;
    }
    
    if (health->average_packet_loss >= thresholds->packet_loss_emergency * 0.6 ||
        health->average_latency >= thresholds->latency_emergency * 0.6) {
        return EMERGENCY_LEVEL_HIGH;
    }
    
    // Medium emergency conditions
    if (health->average_packet_loss >= thresholds->packet_loss_emergency * 0.3 ||
        health->average_latency >= thresholds->latency_emergency * 0.3) {
        return EMERGENCY_LEVEL_MEDIUM;
    }
    
    return EMERGENCY_LEVEL_NONE;
}

// Check cascading failures
static emergency_level_t check_cascading_failures(void) {
    if (!g_emergency_detector_initialized) {
        return EMERGENCY_LEVEL_NONE;
    }
    
    const emergency_thresholds_t* thresholds = &g_emergency_detector.config.emergency_thresholds;
    time_t now = time(NULL);
    
    pthread_mutex_lock(g_emergency_detector.mutex);
    
    // Count recent high-severity incidents
    int recent_high_severity = 0;
    for (int i = 0; i < g_emergency_detector.active_incidents_count; i++) {
        active_incident_t* incident = &g_emergency_detector.active_incidents[i];
        if (incident->severity >= (int)EMERGENCY_LEVEL_HIGH &&
            (now - incident->start_time) < 600) { // 10 minutes
            recent_high_severity++;
        }
    }
    
    if (recent_high_severity >= thresholds->cascading_failure_count) {
        pthread_mutex_unlock(g_emergency_detector.mutex);
        return EMERGENCY_LEVEL_CRITICAL;
    }
    
    // Count failures in the last minute
    int recent_failures = 0;
    time_t cutoff = now - 60; // 1 minute ago
    for (int i = 0; i < g_emergency_detector.failure_records_count; i++) {
        failure_record_t* failure = &g_emergency_detector.recent_failures[i];
        if (failure->timestamp > cutoff) {
            recent_failures++;
        }
    }
    
    pthread_mutex_unlock(g_emergency_detector.mutex);
    
    double failure_rate = (double)recent_failures;
    if (failure_rate >= thresholds->failure_rate_emergency) {
        return EMERGENCY_LEVEL_HIGH;
    }
    
    if (failure_rate >= thresholds->failure_rate_emergency * 0.6) {
        return EMERGENCY_LEVEL_MEDIUM;
    }
    
    return EMERGENCY_LEVEL_NONE;
}

// Check alert-specific emergency conditions
static emergency_level_t check_alert_specific_emergency(notification_type_t alert_type, const char* details_json) {
    if (!g_emergency_detector_initialized || !details_json) {
        return EMERGENCY_LEVEL_NONE;
    }
    
    const emergency_thresholds_t* thresholds = &g_emergency_detector.config.emergency_thresholds;
    
    switch (alert_type) {
        case NOTIFICATION_TYPE_SYSTEM_HEALTH:
            // Parse temperature from JSON details
            if (strstr(details_json, "temperature")) {
                double temperature = 0.0;
                if (sscanf(details_json, "{\"temperature\":%lf", &temperature) == 1) {
                    if (temperature >= thresholds->temperature_emergency) {
                        return EMERGENCY_LEVEL_CRITICAL;
                    }
                    if (temperature >= thresholds->temperature_emergency * 0.8) {
                        return EMERGENCY_LEVEL_HIGH;
                    }
                }
            }
            break;
            
        case NOTIFICATION_TYPE_FAILOVER:
            // Check for multiple rapid failovers
            if (strstr(details_json, "recent_failover_count")) {
                int failover_count = 0;
                if (sscanf(details_json, "{\"recent_failover_count\":%d", &failover_count) == 1) {
                    if (failover_count >= 3) {
                        return EMERGENCY_LEVEL_CRITICAL;
                    }
                    if (failover_count >= 2) {
                        return EMERGENCY_LEVEL_HIGH;
                    }
                }
            }
            break;
            
        case NOTIFICATION_TYPE_DATA_LIMIT:
            // Check for critical data usage
            if (strstr(details_json, "usage_percentage")) {
                double usage_percentage = 0.0;
                if (sscanf(details_json, "{\"usage_percentage\":%lf", &usage_percentage) == 1) {
                    if (usage_percentage >= 95.0) {
                        return EMERGENCY_LEVEL_CRITICAL;
                    }
                    if (usage_percentage >= 90.0) {
                        return EMERGENCY_LEVEL_HIGH;
                    }
                }
            }
            break;
            
        default:
            break;
    }
    
    return EMERGENCY_LEVEL_NONE;
}

// Check temporal emergency patterns
static emergency_level_t check_temporal_emergency(const system_state_t* system_state) {
    if (!system_state || !g_emergency_detector_initialized) {
        return EMERGENCY_LEVEL_NONE;
    }
    
    const emergency_thresholds_t* thresholds = &g_emergency_detector.config.emergency_thresholds;
    time_t now = time(NULL);
    
    // Check for sustained high load
    if (system_state->system_health.cpu_usage > thresholds->cpu_usage_emergency * 0.7 &&
        (now - system_state->system_health.timestamp) > 300) { // 5 minutes
        return EMERGENCY_LEVEL_HIGH;
    }
    
    // Check for rapid state changes
    if (system_state->state_change_count > thresholds->rapid_state_change_threshold) {
        return EMERGENCY_LEVEL_MEDIUM;
    }
    
    return EMERGENCY_LEVEL_NONE;
}

// Detect emergency conditions
emergency_level_t emergency_detector_detect_emergency(const system_state_t* system_state,
                                                    notification_type_t alert_type,
                                                    const char* details_json) {
    if (!g_emergency_detector_initialized || !system_state) {
        return EMERGENCY_LEVEL_NONE;
    }
    
    if (!g_emergency_detector.config.emergency_detection_enabled) {
        return EMERGENCY_LEVEL_NONE;
    }
    
    emergency_level_t max_level = EMERGENCY_LEVEL_NONE;
    
    // Check system health emergencies
    emergency_level_t system_level = check_system_health_emergency(&system_state->system_health);
    if (system_level > max_level) {
        max_level = system_level;
    }
    
    // Check network health emergencies
    emergency_level_t network_level = check_network_health_emergency(&system_state->network_health);
    if (network_level > max_level) {
        max_level = network_level;
    }
    
    // Check cascading failure patterns
    emergency_level_t cascade_level = check_cascading_failures();
    if (cascade_level > max_level) {
        max_level = cascade_level;
    }
    
    // Check alert-specific emergency conditions
    emergency_level_t alert_level = check_alert_specific_emergency(alert_type, details_json);
    if (alert_level > max_level) {
        max_level = alert_level;
    }
    
    // Check temporal emergency patterns
    emergency_level_t temporal_level = check_temporal_emergency(system_state);
    if (temporal_level > max_level) {
        max_level = temporal_level;
    }
    
    // Log emergency detection
    if (max_level > EMERGENCY_LEVEL_NONE) {
        // Note: In a real implementation, this would use the logging system
        printf("EMERGENCY: Level %d detected for alert type %d\n", max_level, alert_type);
    }
    
    return max_level;
}

// Add active incident
static int emergency_detector_add_incident(const active_incident_t* incident) {
    if (!g_emergency_detector_initialized || !incident) {
        return -1;
    }
    
    pthread_mutex_lock(g_emergency_detector.mutex);
    
    if (g_emergency_detector.active_incidents_count >= g_emergency_detector.max_active_incidents) {
        // Remove oldest incident to make room
        for (int i = 0; i < g_emergency_detector.max_active_incidents - 1; i++) {
            g_emergency_detector.active_incidents[i] = g_emergency_detector.active_incidents[i + 1];
        }
        g_emergency_detector.active_incidents_count--;
    }
    
    int index = g_emergency_detector.active_incidents_count;
    g_emergency_detector.active_incidents[index] = *incident;
    g_emergency_detector.active_incidents_count++;
    
    pthread_mutex_unlock(g_emergency_detector.mutex);
    return 0;
}

// Add failure record
static int emergency_detector_add_failure(const failure_record_t* failure) {
    if (!g_emergency_detector_initialized || !failure) {
        return -1;
    }
    
    pthread_mutex_lock(g_emergency_detector.mutex);
    
    if (g_emergency_detector.failure_records_count >= g_emergency_detector.max_failure_records) {
        // Remove oldest failure to make room
        for (int i = 0; i < g_emergency_detector.max_failure_records - 1; i++) {
            g_emergency_detector.recent_failures[i] = g_emergency_detector.recent_failures[i + 1];
        }
        g_emergency_detector.failure_records_count--;
    }
    
    int index = g_emergency_detector.failure_records_count;
    g_emergency_detector.recent_failures[index] = *failure;
    g_emergency_detector.failure_records_count++;
    
    pthread_mutex_unlock(g_emergency_detector.mutex);
    return 0;
}

// Get emergency detector status
static void emergency_detector_get_status(emergency_detector_status_t* status) {
    if (!status || !g_emergency_detector_initialized) return;
    
    pthread_mutex_lock(g_emergency_detector.mutex);
    
    status->enabled = g_emergency_detector.config.emergency_detection_enabled;
    status->active_incidents_count = g_emergency_detector.active_incidents_count;
    status->max_active_incidents = g_emergency_detector.max_active_incidents;
    status->failure_records_count = g_emergency_detector.failure_records_count;
    status->max_failure_records = g_emergency_detector.max_failure_records;
    
    pthread_mutex_unlock(g_emergency_detector.mutex);
}

// Check if emergency detector is initialized
static bool emergency_detector_is_initialized(void) {
    return g_emergency_detector_initialized;
}

// Get emergency detector instance
static emergency_detector_t* emergency_detector_get_instance(void) {
    return g_emergency_detector_initialized ? &g_emergency_detector : NULL;
}
